feat: Implement HTTP spec

MarcoPolo opened the pull request:
A MVP for libp2p HTTP spec https://github.com/libp2p/specs/pull/508. 

This adds an `Host` for doing request-response with HTTP semantics. The libp2phttp.Host contrasts with a "stream host" (basichost or `host.Host` interface). It is capable of doing HTTP over an HTTP transport and/or a stream transport (provided by the stream host).

What works in this MVP:
- `.well-known/libp2p`
- A helper func to give you the best roundtripper given a list of multiaddrs and an optional stream host.
- A way of listening for HTTP on stream transports as well as a native HTTP transport.

My thoughts after implementing the spec is that the spec is a nice minimal set of constraints that are relatively easy to implement. Even making a compliant node using the the stock HTTP library is fairly straightforward.

If this looks good after a review, I would suggest merging this as an experimental feature to allow folks to start using it and getting feedback. I may try to integrate this into the IPNI stack to see how it feels.

Note: This branch targets `marco/gostream` which pulls in https://github.com/libp2p/go-libp2p-gostream into the monorepo.

---

Update Aug 2:
I've removed the libp2p-noise auth from the http spec, and replaced it with this spec focused only on auth: https://github.com/libp2p/specs/pull/564.

I think for the mvp we can merge this without doing any peer id auth, since peer id auth is optional anyways.

BigLep commented:
@MarcoPolo : for the "libp2p and HTTP" effort, what will the testing story be?   I'm not saying that all needs to be in this PR, but I'm curious about where we're aiming to end up.

MarcoPolo commented:
> @MarcoPolo : for the "libp2p and HTTP" effort, what will the testing story be? I'm not saying that all needs to be in this PR, but I'm curious about where we're aiming to end up.

This PR adds basic tests that test end-to-end communication on top of stream transports or an HTTP transport. It tests that a stock HTTP client/server can interoperate. It also adds tests around peer id authentication. It doesn't currently, but will include a test that tests the SNI+TLS flow for peer id authentication, and tests that incorrect auth steps correctly fail (these are noted in before merge list in the PR body).

Once we have multiple implementations of the spec, we will leverage the multidim interop tester to test this as well. It should just work, with no changes needed to the test runner.

Was there a test you had in mind that was missing here? Or something that you feel should be highlighted?

BigLep commented:
> Was there a test you had in mind that was missing here? Or something that you feel should be highlighted?

Thanks for explaining.  I didn't have anything else in mind.  I had missed the tests you described when I took a quick scan earlier.  My bad for not catching those.

Concerning multidimen interop running, we'll also need that for testing version compatibility as well right (even if go-libp2p is the only implementation to support this functionality for awhile)?

MarcoPolo commented:
I've removed the libp2p-noise auth from the http spec, and replaced it with this spec focused only on auth: https://github.com/libp2p/specs/pull/564. I've also removed auth from this PR.

I think for the mvp we can merge this without doing any peer id auth, since peer id auth is optional anyways.

gammazero commented on p2p/http/libp2phttp.go:
```diff
@@ -0,0 +1,525 @@
+
+		ma.ForEach(httpAddrs[0], func(c ma.Component) bool {
+			p := c.Protocol()
+			if p.Code == ma.P_IP4 || p.Code == ma.P_IP6 || p.Code == ma.P_DNS || p.Code == ma.P_DNS4 || p.Code == ma.P_DNS6 {
+				host = c.Value()
+			} else if p.Code == ma.P_TCP || p.Code == ma.P_UDP {
+				port = c.Value()
+			} else if p.Code == ma.P_TLS {
+				useHTTPS = true
+			} else if p.Code == ma.P_SNI {
+				sni = c.Value()
+			}
```
```suggestion
			switch p.Code {
			case ma.P_IP4, ma.P_IP6, ma.P_DNS, ma.P_DNS4, ma.P_DNS6:
				host = c.Value()
			case ma.P_TCP, ma.P_UDP:
				port = c.Value()
			case ma.P_TLS:
				useHTTPS = true
			case ma.P_SNI:
				sni = c.Value()
			}
```

gammazero commented on p2p/http/ping.go:
```diff
@@ -0,0 +1,67 @@
+
+	w.Header().Set("Content-Type", "application/octet-stream")
+	w.Header().Set("Content-Length", strconv.Itoa(pingSize))
+	w.WriteHeader(http.StatusOK)
```
Unnecessary
```suggestion
```

gammazero commented on p2p/http/libp2phttp.go:
```diff
@@ -0,0 +1,525 @@
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	w.WriteHeader(http.StatusOK)
```
[Unnecessary if status is `http.StatusOK`](https://pkg.go.dev/net/http@go1.20.7#ResponseWriter)
```suggestion
```

MarcoPolo replied:
True, although there is no harm in being explicit.

gammazero commented on p2p/http/ping.go:
```diff
@@ -0,0 +1,67 @@
+
+	rBody := [pingSize]byte{}
+	_, err = io.ReadFull(resp.Body, rBody[:])
+	if err != nil {
```
```suggestion
	defer resp.Body.Close()
	if err != nil {
```
[It is the caller's responsibility to close Body](https://pkg.go.dev/net/http@go1.20.7#Response)

gammazero commented on p2p/http/libp2phttp.go:
```diff
@@ -0,0 +1,525 @@
+		bytesRead += n
+		if err == io.EOF {
+			break
+		} else if err != nil {
```
```suggestion
		}
		if err != nil {
```

gammazero reviewed (commented):
This looks great. I plan on using it soon.
Left some trivial suggestions.

MarcoPolo commented:
@gammazero I have a surprise for you: https://github.com/ipni/go-libipni/pull/102

MarcoPolo commented on p2p/http/libp2phttp.go:
```diff
@@ -0,0 +1,675 @@
+	rootHandler      http.ServeMux
+	wk               WellKnownHandler
+	httpRoundTripper *http.Transport
+	recentHTTPAddrs  *lru.Cache[peer.ID, httpAddr]
```
Let's document these and there purpose (both lru cache). Why aren't we using a peer store?

MarcoPolo commented on p2p/security/noise/pb/payload.proto:
```diff
@@ -4,6 +4,8 @@ package pb;
 message NoiseExtensions {
 	repeated bytes webtransport_certhashes = 1;
 	repeated string stream_muxers = 2;
+	optional string SNI = 3;
```
undo this

MarcoPolo commented on p2p/http/libp2phttp.go:
```diff
@@ -0,0 +1,674 @@
+	// lets us know if we've recently connected to an HTTP endpoint and might
+	// have a warm idle connection for it (managed by the underlying HTTP
+	// roundtripper). In some cases, this lets us reuse our existing custom roundtripper (i.e. SNI != host).
+	recentHTTPAddrs *lru.Cache[peer.ID, httpAddr]
```
This should probably be keyed by the addr+scheme+sni combo.

marten-seemann commented on p2p/http/libp2phttp.go:
```diff
@@ -0,0 +1,688 @@
+// HTTPHost is a libp2p host for request/responses with HTTP semantics. This is
+// in contrast to a stream-oriented host like the host.Host interface. Warning,
+// this is experimental. The API will likely change.
+type HTTPHost struct {
```
In analogy with `net/http`, should this be called `Server`?

I'm also wondering if we should get rid of the constructor (`net/http.Server` doesn't have one either). My first attempt at setting up a server was:
```go
host := libp2phttp.HTTPHost{}
host.SetHttpHandler("/website", handler)
log.Fatal(host.Serve())
```
which panicked immediately.

MarcoPolo replied:
> In analogy with net/http, should this be called Server?

This can be used by the client as well (just like the stream host).

> I'm also wondering if we should get rid of the constructor

Not opposed. Should options be public fields as well then?

MarcoPolo replied:
I've made this change. I've removed the constructor entirely and options are set by setting the public fields.

marten-seemann commented on p2p/http/libp2phttp.go:
```diff
@@ -0,0 +1,688 @@
+
+func parseMultiaddr(addr ma.Multiaddr) httpMultiaddr {
+	out := httpMultiaddr{}
+	ma.ForEach(addr, func(c ma.Component) bool {
+		switch c.Protocol().Code {
+		case ma.P_IP4, ma.P_IP6, ma.P_DNS, ma.P_DNS4, ma.P_DNS6:
+			out.host = c.Value()
+		case ma.P_TCP, ma.P_UDP:
+			out.port = c.Value()
+		case ma.P_TLS:
+			out.useHTTPS = true
+		case ma.P_SNI:
+			out.sni = c.Value()
+
+		}
+		return out.host == "" || out.port == "" || !out.useHTTPS || out.sni == ""
+	})
```
Please don't reimplement multiaddr parsing. `manet.ToNetAddr` should provide half of what you need here.

marten-seemann commented on p2p/http/libp2phttp.go:
```diff
@@ -0,0 +1,688 @@
+}
+
+// SetHttpHandler sets the HTTP handler for a given protocol. Automatically
+// manages the .well-known/libp2p mapping.
```
Add an example here. It's not clear if the protocol ID contains a leading / or not, for example.

MarcoPolo replied:
It shouldn't matter. But I agree that an example would be good.

marten-seemann reviewed (commented):
Lots of my comments come from looking at the GoDoc (`godoc -http=:6060`), and putting myself into the shoes of a user of this package.

Should we split up libp2phttp.go into two files, one for the client and one for the server?

---

The following code doesn't work:
```go
host, err := libp2phttp.New(libp2phttp.ListenAddrs([]ma.Multiaddr{ma.StringCast("/ip4/127.0.0.1/tcp/8080/http")}))
if err != nil {
	log.Fatal(err)
}
host.SetHttpHandler("/website", http.FileServer(http.Dir("/tmp")))
log.Fatal(host.Serve())
```

Haven't debugged it yet. It creates the well-known handler, but http://localhost:8080/website/ returns a 404.

MarcoPolo commented on p2p/http/ping.go:
```diff
@@ -0,0 +1,67 @@
+	if err != nil {
+		return err
+	}
+	req, err := http.NewRequest("POST", "/", bytes.NewReader(body[:]))
```
This is namespaced.

MarcoPolo commented:
> Lots of my comments come from looking at the GoDoc (`godoc -http=:6060`), and putting myself into the shoes of a user of this package.

Nice! Thanks!


> The following code doesn't work:
> 
> ```go
> host, err := libp2phttp.New(libp2phttp.ListenAddrs([]ma.Multiaddr{ma.StringCast("/ip4/127.0.0.1/tcp/8080/http")}))
> if err != nil {
> 	log.Fatal(err)
> }
> host.SetHttpHandler("/website", http.FileServer(http.Dir("/tmp")))
> log.Fatal(host.Serve())
> ```

When you go to "/website" the fileserver is seeing the full path "/website" and you get a 404 because you don't have a `/tmp/website` folder. This is the exact same behavior you get with a servemux (`m.Handle("/website/", http.FileServer(http.Dir("/tmp")))`). 

You should strip the prefix of the path:
```
host.SetHttpHandler("/website", http.StripPrefix("/website", http.FileServer(http.Dir("/tmp"))))
```

Maybe this should be the default behavior for SetHttpHandler?

MarcoPolo commented on p2p/http/libp2phttp.go:
```diff
@@ -0,0 +1,688 @@
+				errCh <- srv.ServeTLS(l, "", "")
+			}()
+		} else {
+			go func() {
+				errCh <- http.Serve(l, &h.rootHandler)
+			}()
```
I'm thinking about how we'd support a proxy (like nginx) that terminating SSL and communicating over HTTP.

Maybe it's better to make that a bit harder to avoid users broadcasting unencrypted HTTP by accident.

MarcoPolo commented on p2p/http/libp2phttp.go:
```diff
@@ -0,0 +1,688 @@
+	}
+	w.Header().Add("Content-Type", "application/json")
+	w.Header().Add("Content-Length", strconv.Itoa(len(mapping)))
+	w.WriteHeader(http.StatusOK)
```
https://github.com/libp2p/go-libp2p/pull/2438#discussion_r1284629402

I guess I'm out voted. I'll change it. For the record, I like the consistency of always writing the header, not just in error/informational cases.

MarcoPolo commented on p2p/http/libp2phttp.go:
```diff
@@ -0,0 +1,688 @@
+// TODOs:
+// - integrate with the conn gater and resource manager
+
+type WellKnownProtocolMeta struct {
```
Using a struct gives us more options for backwards compatibility in the future. If we add more to this besides a path, we don't introduce a breaking api change.

Maybe just ProtocolMeta? It's not configuration for a protocol, it's metadata about where this protocol is mounted on this HTTP server.

MarcoPolo replied:
I've renamed to `ProtocolMeta`

MarcoPolo commented:
I've changed the behavior of SetHTTPHandler to strip the prefix for the handler automatically. I can't think of use cases where you wouldn't want that. You can still avoid it by setting a handler on the underlying servemux directly.

MarcoPolo commented on p2p/http/libp2phttp.go:
```diff
@@ -0,0 +1,688 @@
+}
+
+// StreamHostListen retuns a net.Listener that listens on libp2p streams for HTTP/1.1 messages.
+func StreamHostListen(streamHost host.Host) (net.Listener, error) {
```
I'm hiding this function

Wondertan commented on p2p/http/ping/ping.go:
```diff
@@ -0,0 +1,67 @@
+
+// SendPing send an ping request over HTTP. The provided client should be namespaced to the Ping protocol.
+func SendPing(client http.Client) error {
+	body := [32]byte{}
```
```suggestion
	body := [pingSize]byte{}
```

Wondertan commented on p2p/http/libp2phttp.go:
```diff
@@ -0,0 +1,744 @@
+	return resp, nil
+}
+
+// roundTripperForSpecificHost is an http.RoundTripper targets a specific server. Still reuses the underlying RoundTripper for the requests.
```
```suggestion
// roundTripperForSpecificServer is an http.RoundTripper targets a specific server. Still reuses the underlying RoundTripper for the requests.
```

gammazero commented on p2p/http/libp2phttp.go:
```diff
@@ -0,0 +1,744 @@
+					}
+					errCh <- srv.ServeTLS(l, "", "")
+				}()
+				h.httpTransport.listenAddrs = append(h.httpTransport.listenAddrs, listenAddr)
```
This causes a race with `Addrs() `

If `Serve` is called in a separate goroutine, then there needs to be some synchronization to know when `Addrs()` cannot be called. Or, `h.httpTransport.listenAddrs` needs to be set in some function that is called before calling `Serve`. This worked previously because the constructor served that purpose.

Seems that a constructor is needed. Or, maybe a `Start` function that does this setup and calls `Serve` in a separate goroutine and returns.

gammazero commented on p2p/http/libp2phttp.go:
```diff
@@ -0,0 +1,744 @@
+// http.StripPrefix is called on the handler, so the handler will be unaware of
+// it's prefix path.
+func (h *HTTPHost) SetHTTPHandlerAtPath(p protocol.ID, path string, handler http.Handler) {
+	if path[len(path)-1] != '/' {
```
If an empty `path` is passed in, this will panic. Caller will need to check for empty `path`. Maybe safer to handle the `len(path) == 0` case?

MarcoPolo replied:
Also typo on it’s

gammazero reviewed (approved):
I am currently using this here: https://github.com/ipni/go-libipni/pull/113

This looks great and I cannot wait until it gets merged.

MarcoPolo commented on p2p/http/libp2phttp.go:
```diff
@@ -0,0 +1,754 @@
+// See examples for usage.
+//
+//	Warning, this is experimental. The API will likely change.
+type HTTPHost struct {
```
Should this just be called `Host`? Users will use this as `libp2phttp.Host` so it should be easy to differentiate against the core stream-based `host.Host` type.

gammazero replied:
Yes. Avoids stutter. If everything were defined in a single package, then the type would be `Libp2pHTTPHost`, so `libp2phttp.Host` is appropriate.

marten-seemann commented on p2p/http/libp2phttp.go:
```diff
@@ -0,0 +1,756 @@
+}
+
+// GetPeerMetadata gets a peer's cached protocol metadata from the http host.
+func (h *Host) GetPeerMetadata(server peer.ID) (PeerMeta, bool) {
```
Is there any reason to return the whole map here? Alternative: `SupportsProtocol(peer.ID) (path string, bool)`. It would be nice if we didn't have to make the map an exported type.

MarcoPolo replied:
The reason to export the whole type is to give users flexibility in managing the Host's peerMetadata cache without exposing that cache. This is useful if they want to add out of band knowledge of a peer's metadata. This method along with Set and Remove should let users do anything they need to with this metadata cache.

marten-seemann commented on p2p/http/libp2phttp.go:
```diff
@@ -0,0 +1,756 @@
+}
+
+// RemovePeerMetadata removes a peer's protocol metadata from the http host
+func (h *Host) RemovePeerMetadata(server peer.ID, meta PeerMeta) {
```
Unused `meta`?

marten-seemann commented on p2p/http/libp2phttp.go:
```diff
@@ -0,0 +1,756 @@
+// NewRoundTripper returns an http.RoundTripper that can fulfill and HTTP
+// request to the given server. It may use an HTTP transport or a stream based
+// transport. It is valid to pass an empty server.ID.
+func (h *Host) NewRoundTripper(server peer.AddrInfo, opts ...RoundTripperOption) (http.RoundTripper, error) {
```
I'm struggling with this abstraction. An `http.RoundTripper` is supposed to be able to issue requests to any HTTP host. This is more than just a theoretical concern: The `http.Client` has a `Transport` field that takes an `http.RoundTripper`, and would then be able to follow redirects. If we return a `RoundTripper` here that's scoped a single host, it won't be possible to redirect from one libp2p+HTTP host to another.

I don't have a good solution for this now, I'll have to think about the correct abstraction more.

MarcoPolo replied:
> it won't be possible to redirect from one libp2p+HTTP host to another.

I agree this is an interesting use case we should try to support.

The thing I like about this api is that a user can pass in many multiaddrs and the best transport is chosen for them. There isn't anywhere else we could do this besides when we create a round tripper. But, thinking more about this, this doesn't have to be a property of the core round tripper. We can have a round tripper on top of some core round tripper that can do this logic. It would be great if the core round tripper could, as you say, fulfill any HTTP request regardless of the transport.

Here's one idea how to do this:

1. Change this function to accept no parameters. It just returns an http.RoundTripper.
2. This roundtripper can fulfill HTTP requests to the following URIs:
  a. `http://` and `https://`, just like the http.DefaultTransport does.
  b. `multiaddr:<peer-multiaddr>`. Which would be a new URI scheme for multiaddrs. The value after `multiaddr:` is simply the string encoded version of the multiaddr. The multiaddr can include a peer's transport information, or could include only the peer id `/p2p/<peer-id>` and rely on discovery or out-of-band knowledge of a peer's addresses. Note this multiaddr can also be an http transport multiaddr.

This would work but would lead to us having to define the [httppath component](https://github.com/libp2p/specs/issues/554) since we'll need a way to reference the HTTP path we want to use in the multiaddr.

We would probably also want to register the `multiaddr` uri scheme in IANA (but could defer).

Here's a branch of this idea: https://github.com/libp2p/go-libp2p/compare/marco/http...marco/newroundtripper?expand=1. It includes an httppath implementation, but we should standardize that (not use the private area) if we go that route. It also has a test demonstrating the redirect to other host (using multiaddrs).

MarcoPolo replied:
For now let’s remove the NewRoundtripper and only expose the NamespacedRoundtripper and NamespacedClient. We can then add the NewRoundTripper later. 

The only thing we loose is redirects across hosts. I thinks that’s fine to omit for this initial version. And we have plan above on how to add it.

MarcoPolo replied:
I've renamed this to `NewConstrainedRoundTripper`, I think that will unblock us here. In a future version we can introduce the `NewRoundTripper` that accepts a multiaddr uri and can handle redirects across hosts.

MarcoPolo commented on p2p/http/libp2phttp.go:
```diff
@@ -0,0 +1,756 @@
+// - integrate with the conn gater and resource manager
+
+// ProtocolMeta is metadata about a protocol.
+type ProtocolMeta struct {
```
Config implies that this is a configurable, but this is metadata about a peer's supported protocols.

MarcoPolo commented on p2p/http/libp2phttp.go:
```diff
@@ -0,0 +1,756 @@
+	Path string `json:"path"`
+}
+
+type PeerMeta map[protocol.ID]ProtocolMeta
```
No strong preference here, but I do prefer that `PeerMeta` is consistent with `ProtocolMeta`.

MarcoPolo commented on p2p/http/libp2phttp.go:
```diff
@@ -0,0 +1,756 @@
+	// ServeInsecureHTTP indicates if the server should serve unencrypted HTTP requests over TCP.
+	ServeInsecureHTTP bool
+	// ServeMux is the http.ServeMux used by the server to serve requests
+	ServeMux http.ServeMux
```
No it will not use http.DefaultServeMux. To me that's a default for the standard library, and I think it would be confusing if we referenced it here. Users would create a new host that automatically exposes things they weren't expecting.

The reason this is public is to allow a user to set a custom servemux, or to manually manage the mapping outside of `SetHTTPHandler`. I'll expand the comment to mention this.

MarcoPolo replied:
This should probably be a pointer if we want to let users set a custom servemux

MarcoPolo commented on p2p/http/libp2phttp.go:
```diff
@@ -0,0 +1,756 @@
+	ServeMux http.ServeMux
+
+	// DefaultClientRoundTripper is the default http.RoundTripper for clients
+	DefaultClientRoundTripper *http.Transport
```
Documented

MarcoPolo commented on p2p/http/libp2phttp.go:
```diff
@@ -0,0 +1,756 @@
+const peerMetadataLimit = 8 << 10 // 8KB
+const peerMetadataLRUSize = 256   // How many different peer's metadata to keep in our LRU cache
+
+// TODOs:
+// - integrate with the conn gater and resource manager
```
https://github.com/libp2p/go-libp2p/issues/2511

marten-seemann commented on p2p/http/libp2phttp.go:
```diff
@@ -0,0 +1,790 @@
+	// HTTP requests over TCP.
+	InsecureAllowHTTP bool
+	// ServeMux is the http.ServeMux used by the server to serve requests. If nil,
+	// new serve mux will be allocated. Users may manually add handlers to this
```
```suggestion
	// a new serve mux will be created. Users may manually add handlers to this
```

marten-seemann commented on p2p/http/libp2phttp.go:
```diff
@@ -0,0 +1,790 @@
+
+	// WellKnownHandler is the http handler for the `.well-known/libp2p`
+	// resource. It is responsible for sharing this node's protocol metadata
+	// with other nodes.
```
Mention that users only need to care about it if they set their own `ServeMux`, and that it's handled automatically otherwise (if I understand correctly).

marten-seemann commented on p2p/http/libp2phttp.go:
```diff
@@ -0,0 +1,790 @@
+		}
+	}
+
+	err := func() error {
```
Maybe cleaner: Make this a separate method that returns the listen addrs and an error.

MarcoPolo replied:
This also adds to the listeners. Maybe I'll have this accept a pointer of http transport so it's clear what is getting mutated.

MarcoPolo replied:
Actually moved this out into a separate method so it was clear what is in scope vs not.

marten-seemann commented on p2p/http/libp2phttp.go:
```diff
@@ -0,0 +1,790 @@
+					scheme = "https"
+				}
+				listenAddr = ma.StringCast(fmt.Sprintf("/ip4/%s/tcp/%s/%s", host, port, scheme))
+
```
```suggestion
```

marten-seemann commented on p2p/http/libp2phttp.go:
```diff
@@ -0,0 +1,790 @@
+	httpHost *Host
+}
+
+type streamReadCloser struct {
```
Can you add a comment why you're introducing this type (as opposed to just using the underlying `network.Stream`).

marten-seemann commented on p2p/http/libp2phttp.go:
```diff
@@ -0,0 +1,790 @@
+		return nil, err
+	}
+
+	go func() {
+		defer s.CloseWrite()
+		r.Write(s)
+		if r.Body != nil {
+			r.Body.Close()
+		}
+	}()
+
+	resp, err := http.ReadResponse(bufio.NewReader(s), r)
+	if err != nil {
+		return nil, err
+	}
```
You're ignoring the request context here. If this causes too much complexity for the first iteration, feel free to ignore, but please add a TODO to the code.

MarcoPolo replied:
https://github.com/libp2p/go-libp2p/issues/2533

marten-seemann commented on p2p/http/libp2phttp.go:
```diff
@@ -0,0 +1,790 @@
+	// connections
+}
+
+type namespacedRoundTripper struct {
```
Add a doc comment?

marten-seemann commented on p2p/http/libp2phttp.go:
```diff
@@ -0,0 +1,790 @@
+
+// NewConstrainedRoundTripper returns an http.RoundTripper that can fulfill and HTTP
+// request to the given server. It may use an HTTP transport or a stream based
+// transport. It is valid to pass an empty server.ID.
```
Describe the logic that's used to use a stream based vs. an HTTP transport?

marten-seemann commented on p2p/http/libp2phttp.go:
```diff
@@ -0,0 +1,790 @@
+// NewConstrainedRoundTripper returns an http.RoundTripper that can fulfill and HTTP
+// request to the given server. It may use an HTTP transport or a stream based
+// transport. It is valid to pass an empty server.ID.
+func (h *Host) NewConstrainedRoundTripper(server peer.AddrInfo, opts ...RoundTripperOption) (http.RoundTripper, error) {
```
This function is a bit problematic, as it dials a new connection (which might take an extended period of time), without providing a way to cancel that. One option would be to add a context, but that would be a bit surprising for a function that look like a constructor. Another option would be to defer dialing.

marten-seemann commented on p2p/http/libp2phttp.go:
```diff
@@ -0,0 +1,790 @@
+		return nil, fmt.Errorf("unexpected status code: %d", resp.StatusCode)
+	}
+
+	body := [peerMetadataLimit]byte{}
+	bytesRead := 0
+	for {
+		n, err := resp.Body.Read(body[bytesRead:])
+		bytesRead += n
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return nil, err
+		}
+		if bytesRead >= peerMetadataLimit {
+			return nil, fmt.Errorf("peer metadata too large")
+		}
+	}
```
Consider using an `io.LimitedReader`. Avoids allocating 8 kB, which won't be needed in the vast majority of cases.

marten-seemann commented on p2p/http/libp2phttp.go:
```diff
@@ -0,0 +1,790 @@
+	}
+
+	meta := PeerMeta{}
+	json.Unmarshal(body[:bytesRead], &meta)
```
Ignored error return value?

gammazero commented on p2p/http/libp2phttp.go:
```diff
@@ -0,0 +1,804 @@
+	meta := PeerMeta{}
+	err = json.NewDecoder(&io.LimitedReader{
+		R: resp.Body,
+		N: 8 << 10,
```
```suggestion
		N: peerMetadataLimit,
```

marten-seemann reviewed (approved):
Ship it!
